### PR TITLE
foomatic-db ppd files

### PR DESCRIPTION
**Pull request by Yarny0**
###### Motivation for this change

OpenPrinting provides about 10,000 ppd files for printers, which currently need to be downloaded and installed manually for NixOS.  It would be nice to have them accessible as a package so they can be used declaratively (e.g. via `config.hardware.printers.ensurePrinters`).  This pull-request presents the necessary packages that generate and provide those ppd files.  In the course of doing so, it also adds a new setup hook that helps patching executable paths in pdd files.  The setup hook is also useful for other packages that generate ppd files -- this pull-request adds the new setup hook to those packages as well.  To keep everything clearly laid out, the changes are split into several commits.

###### Summary of the changes

Each of the following points is accomplished by one or more commits.  The commit messages contain more information on the changes and some reasoning on implementation details.

* New setup hook `patch-ppd-files`

  Many ppd files (from `foomatic-db*` as well as from other packages) contain calls to executables (e.g. `gs` or `perl`).  If these are called without absolute path, they work only if the command is in CUPS' PATH.  Other ppd files contain absolute paths (e.g. `/usr/bin/gs`) and require patching to work with NixOS at all.  Some packages already patch their ppd files to replace these commands with absolute paths into the nix store.  The setup hook `patch-ppd-files` tries to unify these efforts in a most robust way.

* Use `patch-ppd-files` on existing packages

  Some packages in nixpkgs patch executable paths in ppd files manually, e.g. with `sed`.  This pull requests replaces these efforts with the new setup hook `patch-ppd-files`.  Since I can't test those packages with real hardware (due to lack of printers), I decided to touch only those packages that already do patching so I can easily verify that results are the same and minimize the probability of breakage.

  Affected packages are

    * `cups-drv-rastertosag-gdi`
        * maintained by myself
        * the build recipe didn't patch paths before
        * also added gzip compression to ppd files
          (separate commit)
        * tested with real printer (see below)

    * `cups-kyocera`
        * maintained by @vanzef
        * just replaced patching logic with new setup hook
        * the ppd files are identical (up to package's hash)

    * `samsung-unified-linux-driver`
        * maintained by @jokogr
        * replaced patching logic with new setup hook
        * also added a missing executable's
          name to the patching logic

  The package `cups-kyodialog3` would also be a candidate, but it fails to build so I cannot verify the result.

* Introduce packages containing the Foomatic database

  The packages `foomatic-db` and `foomatic-db-nonfree` provide XML files containing the so-called Foomatic database, which contains structured information about all printers listed on https://openprinting.org/printers .  This database can be used to generate ppd files (see below).  Besides, these packages have about 6,300 ppd files on their own.  These ppd files are patched with the setup hook (see above).

* Introduce package `foomatic-db-engine`

  The package `foomatic-db-engine` contains a perl module and some command line scripts that can be used to query the Fooomatic database XML files.  The package will be used to extract ppd files.

* Introduce package `foomatic-db-ppds` containing all ppd files from the Foomatic database

  This package does not fetch a source file.  Instead, it uses the Foomatic database packages and the Foomatic database engine (see above) to generate all ppd files known to the database.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS (on `x86-64`)
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Tests with hardware (note that these tests were performed with stable NixOS releases as my access to printers is limited):

- [x] tested `foomatic-db-ppds/NEC/NEC-PinWriter_P6-necp6.ppd.gz` with printer "NEC PinWriter P5XL" (in NixOS 20.09, 21.05)
- [x] tested `foomatic-db-ppds/Xerox/Xerox-WorkCentre_7345-pxlcolor.ppd.gz` with printer "Xerox WorkCentre 7835 v1 Multifunction System" (in NIixOS 21.05)
- [x] tested `foomatic-db-ppds/Samsung/Samsung-ML-1440-gdi.ppd.gz` with printer "Samsung ML-1440" (in NixOS 21.05)
- [x] tested `foomatic-db-ppds/Samsung/Samsung-ML-371x-Postscript-Samsung.ppd.gz` with printer "Samsung ML-3710ND" (in NixOS 21.05)
- [x] tested `foomatic-db/Samsung/PS/Samsung_ML-371x_Series.ppd.gz` with printer "Samsung ML-3710ND" (in NixOS 21.05)
- [x] tested package `cups-drv-rastertosag-gdi` with printer `Philips 6050W 4-in-1` (in NixOS 20.09, 21.05)

###### Further comments

* Notifying `cups-kyocera` maintainer @vanzef and `samsung-unified-linux-driver` maintainer @jokogr:  This pull request replaces the ppd file patching code in your packages with the new `patch-ppd-files` hook.  As far as I can judge, the result is the same but you might want to test these packages or otherwise comment on the changes.  If you can spare the time, I would also appreciate a review of the pull request as a whole.

* Overview of similar packages in other distributions:  https://repology.org/projects/?search=foomatic-db

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 133537` at 123e9175 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built successfully:</summary>
  <ul>
    <li>cups-drv-rastertosag-gdi</li>
    <li>foomatic-db-engine</li>
    <li>samsung-unified-linux-driver (samsung-unified-linux-driver_4_01_17)</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/misc/cups/drivers/samsung/4.01.17.nix:42:3:
```
   |
42 |   installPhase = ''
   |   ^
```
</li>
  </ul>
</details>

File: pkgs/misc/cups/drivers/foomatic-db/default.nix
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/foomatic-db/default.nix:75]:
```suggestion
    echo 'compressing ppd files'
```

File: pkgs/misc/cups/drivers/foomatic-db/default.nix
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/foomatic-db/default.nix:49]:
```suggestion
  configureFlags = [ "--disable-gzip-ppds" ];
```

File: pkgs/misc/cups/drivers/foomatic-db-ppds/default.nix
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/foomatic-db-ppds/default.nix:106]:
```suggestion
    echo 'compressing ppd files'
```

File: pkgs/misc/cups/drivers/foomatic-db-ppds/default.nix
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/foomatic-db-ppds/default.nix:52]:
```suggestion
stdenv.mkDerivation {
  pname = "foomatic-db-ppds";
```

File: pkgs/misc/cups/drivers/foomatic-db-ppds/default.nix
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/foomatic-db-ppds/default.nix:28]:
```suggestion
      echo 'export FOOMATICDB="${placeholder "out"}/share/foomatic"' \
```

File: pkgs/misc/cups/drivers/foomatic-db-ppds/default.nix
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/foomatic-db-ppds/default.nix:22]:
```suggestion
}:

let
  foomatic-db-packages = [ foomatic-db foomatic-db-nonfree ];

  foomatic-db-combined = buildEnv {
```

File: pkgs/misc/cups/drivers/foomatic-db-nonfree/default.nix
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/foomatic-db-nonfree/default.nix:56]:
```suggestion
    echo 'compressing ppd files'
```

File: pkgs/misc/cups/drivers/foomatic-db-nonfree/default.nix
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/foomatic-db-nonfree/default.nix:48]:
```suggestion
    ln -vs "${placeholder "out"}/share/foomatic/db/source/PPD"  \
      "${placeholder "out"}/share/cups/model/foomatic-db-nonfree"
```

File: pkgs/misc/cups/drivers/foomatic-db-engine/default.nix
Comment (SuperSandro2000) [inline pkgs/misc/cups/drivers/foomatic-db-engine/default.nix:49]:
```suggestion
  preInstall = ''
    mkdir -vp "${placeholder "out"}/lib/cups"
  '';
```

File: pkgs/build-support/setup-hooks/patch-ppd-files/patch-ppd-hook.sh
Comment (SuperSandro2000) [inline pkgs/build-support/setup-hooks/patch-ppd-files/patch-ppd-hook.sh:155]:
```suggestion
        echo 'patchPpdFileCommands: %s ppd file(s) patched with '"'%s'"''  \
```

File: pkgs/build-support/setup-hooks/patch-ppd-files/patch-ppd-hook.sh
Comment (SuperSandro2000) [inline pkgs/build-support/setup-hooks/patch-ppd-files/patch-ppd-hook.sh:66]:
```suggestion
    local bin binnew binold binoldgrep cupspath path ppdroot ppdrootprefix
```

**Review by SuperSandro2000 — COMMENTED** [review]
You get the gist. I did not comment everything.
Also we should probably remove the -v from most commands to make the build less noisy.

**Comment by SuperSandro2000** [comment]
> I'm not sure if all echo implementations that might come into play here can handle situations where parts of the string look like a command line option.

Luckily that does not matter because always the bash from the stdenv is used.

**Comment by Yarny0** [comment]
Thanks @SuperSandro2000.
The branch is updated, all shell script messages are now performed with `echo`.  It is also rebased on top of current `nixos-unstable` commit.
I don't expect any functional changes or breakage due to the update, as it only affects diagnostic messages.  However, to rule out typos I rebuilt all affected packages (on NixOS unstable) and quickly tested the package again, with `foomatic-db-ppds/NEC/NEC-PinWriter_P6-necp6.ppd.gz` with printer "NEC PinWriter P5XL" (in NixOS 21.05), and the printer still prints as expected.

**Comment by Yarny0** [comment]
Another update of the pull request's branch as I just realized https://github.com/OpenPrinting/foomatic-db received some commits in August and September.
They added xml/ppd files for five Ricoh printers:  https://github.com/OpenPrinting/foomatic-db/commit/13e1f23b4dae7987921b89a07fd5ddb0ae5c7a51#diff-91c5b46dc84a94604a4e4d0caed9bf85590a2eddbb12d2e8dc80badf324a9dfb
All packages in the pull request at hand still build fine with the update.  The new ppd files do not contain any new commands that would require an update of the list of commands to be patched.

File: pkgs/build-support/setup-hooks/patch-ppd-files/patch-ppd-hook.sh
Comment (SuperSandro2000) [reply pkgs/build-support/setup-hooks/patch-ppd-files/patch-ppd-hook.sh:60]: Right, if this only exists in the sandbox then this is fine.

**Comment by SuperSandro2000** [comment]
@ofborg eval
@ofborg build

**Comment by Yarny0** [comment]
Branch updated:

* updated `foomatic-db` to version `2022-01-27`
* rebased onto fresh `nixos-unstable`

**Comment by Yarny0** [comment]
Pushed another update:

* updated `foomatic-db` to version `2022-03-27`
* rebased onto current `nixos-unstable` (e10da1c7f542515b609f8dfbcf788f3d85b14936)

All affected packages still build fine.

Tested `foomatic-db-ppds/Samsung/Samsung-ML-371x-Postscript-Samsung.ppd.gz` with printer "Samsung ML-3710ND" (with current NixOS unstable and with current NixOS 21.11).

Is there anything that can be done to help this pull request getting merged into `master`?

**Comment by Yarny0** [comment]
Another update of the branch.

* Rebase onto current `nixos-unstable`, i.e., 93c57a988470c1948976b1bb70abbd5855c5b810 .
* Moved the string "unstable" from the `pname` to the `version` attributes, following 01ec4349f20157dd3bbcc1f3a540eb051c47be50 .
* All new packages still refer to the current version's sources.
* All new/affected packages still build fine, up to one package: The unfree package `cups-kyocera` fails to download it's source archive; the server apparently no longer provides the file.  Notifying maintainer: @vanzef .  I'm still retaining the `cups-kyocera` related commits as they worked fine before and will likely work fine as soon as the package's source archive is available again.

**Comment by Yarny0** [comment]
Updated and rebased the branch.  The branch is now using the newest version of `foomatic-db`, released yesterday.  The release added a couple of RICOH printers and an epson printer.
                                                                                                                                    
All packages still build fine, and the automated test for the new `patch-ppd-files` setup hook still passes.                        
                                                                                                                                    
It would really be nice to have this pull request finally be merged into master.

**Comment by Yarny0** [comment]
Another rebase, including the newest version of `foomatic-db`.  Everything still builds fine.

File: pkgs/misc/cups/drivers/foomatic-db-ppds/foomatic-ppdfiles.pl
Comment (doronbehar) [inline pkgs/misc/cups/drivers/foomatic-db-ppds/foomatic-ppdfiles.pl:9]: Have you considered contributing this script upstream? With or without this script part of `foomatic-db-engine`, I envision that `foomatic-db-ppds` could be removed, and you'll run that script in `foomatic-db{,-nonfree}` (perhaps conditionally) - making these packages include the ppd files from the databases they came with.

File: pkgs/misc/cups/drivers/foomatic-db-ppds/default.nix
Comment (doronbehar) [inline pkgs/misc/cups/drivers/foomatic-db-ppds/default.nix:24]: If we'll leave things as they are regarding this new package, I think it'd be nice to add a comment like this (assuming I understood what's going on correctly):

```suggestion
    # `foomatic-db-combined` is a nativeBuildInput of `foomatic-db-ppds`. Thanks to this
    # setup hook, our `foomatic-ppdfiles.pl` script knows where to find the database to
    # generate ppd files from.
```

**Review by doronbehar — COMMENTED** [review]
Everything is very well written, and commit messages are very detailed! Thanks for the effort to contribute.

File: pkgs/misc/cups/drivers/cups-drv-rastertosag-gdi/default.nix
Comment (Yarny0) [reply pkgs/misc/cups/drivers/cups-drv-rastertosag-gdi/default.nix:41]: Yeah, that sounds reasonable.  I added a hint in the release notes and will post a warning in discourse after it got merged.

File: pkgs/top-level/all-packages.nix
Comment (Yarny0) [reply pkgs/top-level/all-packages.nix]: I agree.  Adding `Hook` in hooks' names seems helpful in general.  Done.

File: pkgs/misc/cups/drivers/samsung/4.01.17.nix
Comment (Yarny0) [reply pkgs/misc/cups/drivers/samsung/4.01.17.nix:85]: I usually try to separate commits if their changes are conceptually different, to facilitate bisection if a bug is introduced.  Here, the first commit should not affect the package's output at all (besides its hash), while the second commit intentionally alters the output, as it fixes an oversight that was already present for a long time.  I think the separation into two commits is justified here, what do you think?

File: pkgs/misc/cups/drivers/foomatic-db/default.nix
Comment (Yarny0) [reply pkgs/misc/cups/drivers/foomatic-db/default.nix]: Yes, that's the idea of this line.  I expanded the line to print an error message.

File: pkgs/misc/cups/drivers/foomatic-db-ppds/foomatic-ppdfiles.pl
Comment (Yarny0) [reply pkgs/misc/cups/drivers/foomatic-db-ppds/foomatic-ppdfiles.pl]: Thanks for bringing up this idea.  The `foomatic-db-engine` maintainer [answered promptly](https://github.com/OpenPrinting/foomatic-db-engine/pull/8#issuecomment-1296183602) and brought another command to my attention:  I didn't spot `foomatic-compiledb` earlier, but it is doing exactly what I tried to accomplish with my tiny script.  Up to the subdirectory structure, the result is the same, so we don't have to patch `foomatic-db-engine` at all.

Note: I also upgraded the `foomatic-db-engine` from 4.0 to 4.1 while preparing the patch.  With this update, the final `foomatic-db-ppds` package "lost" about 1.000 PPD files, but apparently these files were duplicates of files in `foomatic-db`, so we don't really loose anything.

Regarding the ppd export directly in the database packages: Inspired by your idea, I restructured everything to make this work, but alas, `foomatic-db-nonfree` isn't a "database" by itself, but just a kind of extension to the free `foomatic-db`.  The free database would install and export everything alright, but the nonfree version would need to be combined with the free one after installation, only to generate the free ppd files again.  The result of this restructuring would produce working packages, but the build script of the nonfree version would work very differently (and more complicated) compared to the free version, leading to a quite awkward structure overall.  Finally, I came to the conclusion that the current design is more understandable and maintainable, so I suggest to keep it.

File: pkgs/misc/cups/drivers/foomatic-db-ppds/default.nix
Comment (Yarny0) [reply pkgs/misc/cups/drivers/foomatic-db-ppds/default.nix:27]: Yes, I appended a corresponding comment.

File: pkgs/misc/cups/drivers/samsung/4.01.17.nix
Comment (doronbehar) [reply pkgs/misc/cups/drivers/samsung/4.01.17.nix:85]: > I usually try to separate commits if their changes are conceptually different, to facilitate bisection if a bug is introduced. [...] I think the separation into two commits is justified here, what do you think?

I agree with this justification for separating / merging commits. Still I would have merged these two commits because I wouldn't have had the patience to vouch for the previous change to not change the resulting ppd files, and only change the inputs. But that's me, you are much more thorough then me! :)

File: pkgs/misc/cups/drivers/foomatic-db-engine/default.nix
Comment (doronbehar) [inline pkgs/misc/cups/drivers/foomatic-db-engine/default.nix:39]: What do you mean by "improve something"? You noticed the configure script looks for these commands, but you didn't understand what's the usage of the configure script / make files with these commands?

File: pkgs/misc/cups/drivers/foomatic-db-ppds/foomatic-ppdfiles.pl
Comment (doronbehar) [reply pkgs/misc/cups/drivers/foomatic-db-ppds/foomatic-ppdfiles.pl]: The current packages' structure is great, I'm glad you switched to `foomatic-compiledb`. Thanks for the effort!

File: pkgs/misc/cups/drivers/foomatic-db-engine/default.nix
Comment (Yarny0) [reply pkgs/misc/cups/drivers/foomatic-db-engine/default.nix:39]: Somehow yes.  If you put `a2ps` into `nativeBuildInputs`, the configure script will say something like

```
checking for a2ps... /nix/store/fhp0m7anlcx4zpzz7qy0kkyilpag5j5n-a2ps-4.14/bin/a2ps
```

For `a2ps` specifically, I couln't find any usage inside the `foomatic-db-engine` sources, so it *might* be some kind of leftover, but I can't really be sure.

This is similar to the list I put in another comment a few lines above: There are also packages that are referenced by the configure script if found, but they will also become a runtime dependency of the resulting `foomatic-db-engine` package in the end (in contrast to `a2ps` and `ghostscript` which will be "used" during the build process, but not be referenced by the result).

The result works fine without any of those packages as build inputs.  On the other hand, the configure script demands certain build inputs (like the `file` utility), so I came to the conclusion that these packages are deliberately treated as optional inputs and can be omitted to avoid bloating the build and the closure.

However, someday someone might be missing a feature that could be unlocked with one of those optional input packages and I hope these comments might point that someone into the right direction and save some debugging time.

Comment (doronbehar) [reply pkgs/misc/cups/drivers/foomatic-db-engine/default.nix:39]: `./configure` scripts looking up for `file` is a known harmless issue happening with all autotools based software, nothing I'd worry about.

OTH, `foomatic-db-engine`'s README explains what is the usage of the `a2ps` and ghostscript and it seems to me, that they can be used during runtime as filters or something like that. It's not a reason to add them to `nativeBuildInputs`. As for the `buildInputs`, it would have been nice to track down what are the new features gained by adding `cups`, `cups-filter` etc.

If I understand correctly, that most of `foomatic-db-engine`'s purpose is to help build `foomatic-db-ppds`, I trust you that most users will likely use the ppds and that the generated ppds are good. Still, perhaps it'd be best to add these inputs conditionally, defaulting to not adding them; and ideally, explain better in a comment above the `lib.optional` lines what is the feature gained by adding these dependencies. Perhaps permalinks to upstream's README is best.

Comment (Yarny0) [reply pkgs/misc/cups/drivers/foomatic-db-engine/default.nix]: Sadly, the README file didn't really help here.  It recommends to have a couple of additional programs installed, but some of those are not really used by the engine package (probably just recommended to make certain PPD files work).  Other dependencies are not mentioned in the README.

I tracked down the references in the perl code: `ghostscript` and `a2ps` are not used at all; they do not alter the output of the package in any way.  The other dependencies expand the functionalities of the scripts `foomatic-{configure,getpjloptions,printjob}`.  From what I understand, these script can be used to modify printer configuration (e.g. add CUPS queues) or submit print jobs, so they are indeed not used in the process of PPD file generation.

I added three switches `withCupsAccess`, `withSocketAccess` and `withSMBAccess` that control the addition of optional dependencies, as you suggested.  I was thinking about the common name `with...Support` that is used elsewhere in nixpkgs, but concluded it might confuse other contributors into thinking that `foomatic-db-engine` or the PPD files it generates cannot be used with cups unless `withCupsSupport` is enabled.  The name `with...Access` makes it more clear that this is only needed if the resulting package is going to be used to *access* one of those entites directly.

Those new switches are the result of my interpretation of the perl code (in particular, the `sysdeps` structure created by `makeDefaults.pm`).  I'm quite sure that access to cups, to network socket and to SMB can be separated the way presented here.  It might even be possible to divide the cups access feature up further in a meaningful way, but from looking at the code, I can't spot an obvious way to do this.

**Comment by doronbehar** [comment]
Result of `nixpkgs-review pr 133537` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>cups-kyocera</li>
  </ul>
</details>
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>cups-drv-rastertosag-gdi</li>
    <li>foomatic-db</li>
    <li>foomatic-db-engine</li>
    <li>foomatic-db-nonfree</li>
    <li>foomatic-db-ppds</li>
    <li>foomatic-db-ppds-withNonfreeDb</li>
    <li>patchPpdFilesHook</li>
    <li>samsung-unified-linux-driver (samsung-unified-linux-driver_4_01_17)</li>
    <li>samsung-unified-linux-driver_1_00_36</li>
    <li>samsung-unified-linux-driver_1_00_37</li>
  </ul>
</details>

**Review by doronbehar — APPROVED** [review]

**Comment by doronbehar** [comment]
`cups-kyocera` failing to build is not an issue of this PR - the URL of the source seems a bit flaky. Not should be a blocker for this well written PR.

File: pkgs/misc/cups/drivers/foomatic-db-engine/default.nix
Comment (doronbehar) [reply pkgs/misc/cups/drivers/foomatic-db-engine/default.nix]: Very good job on `foomatic-db-engine`! Thanks again!

File: pkgs/misc/cups/drivers/cups-drv-rastertosag-gdi/default.nix
Comment (doronbehar) [reply pkgs/misc/cups/drivers/cups-drv-rastertosag-gdi/default.nix:41]: > Yeah, that sounds reasonable. I added a hint in the release notes and will post a warning in discourse after it got merged.

Just a friendly reminder :)

**Pull request merged**
